### ISSUE TITLE
Use -fno-strict-aliasing

### ISSUE THREAD

**Issue by DemiMarie** [dotnet/coreclr#6454]
The JIT includes this code:

```c++
            float f = 0.0;
            int* addr = NULL;
            if (args->TypeGet() == TYP_FLOAT) 
            {
                f = (float) args->gtDblCon.gtDconVal;
                // *(long*) (&f) used instead of *addr because of of strict
                // pointer aliasing optimization. According to the ISO C/C++
                // standard, an optimizer can assume two pointers of
                // non-compatible types do not point to the same memory.
                inst_IV(INS_push, *((int*) (&f)));
                genSinglePush();
                addrReg = 0;
            }
```

I suspect that this is still invalid under strict aliasing, since a float and int pointer can be assumed not to alias.

This is likely pervasive in CoreCLR, so I suggest using `-fno-strict-aliasing` on compilers that support it, instead of trying to fix every case.

**Comment by GSPP** [comment]
Do all supported compilers support that? In any case it might be worth it to fix these issues long-term. This is undefined behavior. It's playing with fire.

**Comment by DemiMarie** [comment]
Clang, GCC, and Intel all do, and real programs (such as the Linux kernel)
rely on it, so it's not going to go away.   Don't know about MSVC.

**Comment by mikedn** [comment]
>  Don't know about MSVC.

MSVC doesn't have such an option but it doesn't need it anyway, unlike other compilers it doesn't stretch the definition of "undefined behavior" to the limit.

**Comment by choikwa** [comment]
Strict-aliasing is a very "defined" behavior. It allows for code motion between types that do not alias and is a well known static alias analysis. Programs that depend on all pointers aliasing are certainly much harder to debug and CoreCLR would not be an exception, judging from the example @DemiMarie posted. For correctness' sake, it would be better to turn it off until all violations are fixed, but doing so would likely have a performance impact on Non-Windows platforms.

**Comment by mikedn** [comment]
> Strict-aliasing is a very "defined" behavior. It allows for code motion between types that do not alias and is a well known static alias analysis. 

Strict-aliasing is not a behavior, it's a set of rules that mandates what is defined behavior and what not. The actual behavior is determined by each compiler. They may blindly trust TBAA or they may choose to ignore TBAA, especially when it's obvious that the pointers do alias like it's the case here.

**Comment by choikwa** [comment]
Yes. LLVM for one does appear to use TBAA at -O3. AFAIK, TBAA is on since -O2 and this means current release build config is a potential ticking time bomb.

**Comment by davidchisnall** [comment]
If anyone wants to address the potential issues from strict aliasing, it's a good idea to add a CI configuration that uses LLVM's type sanitiser.  This will insert dynamic checks that the TBAA rules are followed and warn about unsafe aliasing.
